[Tutorial] How to create ROOT Dictionary for a custom class is a stand-alone C++ project

atolosadelgado opened the pull request:
Hi,

# This Pull request: 

This PR comes from this [forum thread](https://root-forum.cern.ch/t/use-root-io-as-library-in-a-c-stand-alone-project/55539/5). This tutorial is a minimal working example about how to create a ROOT dictionary for a custom class, and its use for writing and reading a TTree.

Further changes may be needed before full integration into ROOT project.

## Changes or fixes:

New folder added, `tutorials/tree/dictionary`

## Checklist:

- [x] tested changes locally
- [x] updated the docs (if necessary)

jalopezg-git commented:
Many thanks, @atolosadelgado!  This will be reviewed soon!

jalopezg-git commented on tutorials/tree/dictionary/CMakeLists.txt:
```diff
@@ -0,0 +1,55 @@
+# CMakeLists.txt for event package. It creates a library with dictionary and a main program
+cmake_minimum_required(VERSION 3.9 FATAL_ERROR)
+
+project(customTreeExample)
```
```suggestion
project(treeUsingCustomClass)
```

jalopezg-git commented on tutorials/tree/dictionary/README.md:
```diff
@@ -0,0 +1,42 @@
+How to read and write custom classes in TTrees
+==============================================
```
```suggestion
How to read and write custom classes in TTree
=============================================
```

jalopezg-git commented on tutorials/tree/dictionary/README.md:
````diff
@@ -0,0 +1,42 @@
+==============================================
+
+# Starting steps
+
+1. Build the project, it requires ROOT
+
+```
+cmake -B build -S . 
+cmake --build build 
+```
+
+2. Run the executable
+
+```
+./build/treeExample
+```
+
+3. Inspect the output rootfile, and the source code. 
+
````
In general, let's also briefly describe here the steps required to read/write user-defined classes, i.e. why ROOT relies on dictionaries for class introspection, how to generate one, etc.

jalopezg-git commented on tutorials/tree/dictionary/README.md:
```diff
@@ -0,0 +1,42 @@
+
+# Files
+
+The TTree can be seen as a collection of objects (branches), with a number of attributes (leaves). The name of the branch corresponds to the name of the instantianated object. The name of the leaves corresponds to the name of the attributes.
```
```suggestion
The TTree can be seen as a collection of objects (branches), with a number of attributes (leaves). The name of the branch corresponds to the name of the instantiated object. The name of the leaves corresponds to the name of the attributes.
```

jalopezg-git commented on tutorials/tree/dictionary/README.md:
```diff
@@ -0,0 +1,42 @@
+
+The TTree can be seen as a collection of objects (branches), with a number of attributes (leaves). The name of the branch corresponds to the name of the instantianated object. The name of the leaves corresponds to the name of the attributes.
+
+The attributes of 
```
This sentence is unterminated.

jalopezg-git commented on tutorials/tree/dictionary/README.md:
```diff
@@ -0,0 +1,42 @@
+
+The class that is present in the TTree is declared in the .hpp, and the methods defined in the .cpp. 
+
+The linkdef file contains some instructions for ROOT, to specify which classes will require a dictionary:
```
We have not introduced the term "dictionary" before.

jalopezg-git commented on tutorials/tree/dictionary/data2Tree.cpp:
```diff
@@ -0,0 +1,19 @@
+   energy = 0;
+   detectorID = 0;
+   correlatedDetectors_v.clear();
+   return;
```
C/C++ functions are guaranteed to return anyway when execution reaches the end of their body.
```suggestion
```

jalopezg-git commented on tutorials/tree/dictionary/data2Tree.cpp:
```diff
@@ -0,0 +1,19 @@
+// The class method was declared in the corresponding .hpp file
+
+#include "data2Tree.hpp"
+#include <iostream>
```
This does not seem to be used below.
```suggestion
```

jalopezg-git commented on tutorials/tree/dictionary/main.cpp:
```diff
@@ -0,0 +1,20 @@
+// # Copyright (c) 2018 Alvaro Tolosa. All rights reserved.		 #
+// ##########################################################################
+
+#include "readTree.cpp"
+#include "writeTree.cpp"
```
`#include`ing files that provide anything different from templated code, `inline` functions, or any other entities that are safe including from many translation units is, in general, frowned upon.

I would suggest instead to either:
- Directly include the code here (obviously removing the include guards and duplicated `#include` lines), or...
- Separately compile (as different translation units) each of those files and then link the resulting `.o` object files

jalopezg-git commented on tutorials/tree/dictionary/main.cpp:
```diff
@@ -0,0 +1,20 @@
+   std::cerr << "Starting readTree()... Done! " << std::endl;
+
+   return 0;
+}
```
Missing `\n` at the end of last line.

jalopezg-git commented on tutorials/tree/dictionary/writeTree.cpp:
```diff
@@ -0,0 +1,61 @@
+
+void writeTree()
+{
+
```
```suggestion
```

jalopezg-git commented on tutorials/tree/dictionary/writeTree.cpp:
```diff
@@ -0,0 +1,61 @@
+void writeTree()
+{
+
+   TFile *ofile = TFile::Open("testFile.root", "recreate");
```
I would use a `std::unique_ptr` instead here (see https://root.cern/manual/trees/).  I think it's more inline with modern C++ guidelines.

jalopezg-git commented on tutorials/tree/dictionary/writeTree.cpp:
```diff
@@ -0,0 +1,61 @@
+      std::cerr << " File not found or already exists" << std::endl;
+      return;
+   }
+   TTree *myTree = new TTree("myTree", "");
```
Same here for the `std::unique_ptr`.

atolosadelgado replied:
ROOT segfaults if I replace
`TTree *myTree = new TTree("myTree", "");`
by
`std::unique_ptr<TTree> myTree = std::make_unique<TTree>("myTree", "");`

This is the backtrace:
```
 *** Break *** segmentation violation



===========================================================
There was a crash.
This is the entire stack trace of all threads:
===========================================================
#0  0x00007fd5aea6a5db in waitpid () from /lib64/libc.so.6
#1  0x00007fd5ae9ccce7 in do_system () from /lib64/libc.so.6
#2  0x00007fd5b3725563 in TUnixSystem::StackTrace() () from /usr/local/root_6.26.06/lib/libCore.so
#3  0x00007fd5b3727f95 in TUnixSystem::DispatchSignals(ESignals) () from /usr/local/root_6.26.06/lib/libCore.so
#4  <signal handler called>
#5  0x000000000040465a in std::default_delete<TTree>::operator()(TTree*) const ()
#6  0x0000000000404317 in std::unique_ptr<TTree, std::default_delete<TTree> >::~unique_ptr() ()
#7  0x0000000000403e45 in writeTree() ()
#8  0x0000000000402e77 in main ()
===========================================================


The lines below might hint at the cause of the crash.
You may get help by asking at the ROOT forum https://root.cern/forum
Only if you are really convinced it is a bug in ROOT then please submit a
report at https://root.cern/bugs Please post the ENTIRE stack trace
from above as an attachment in addition to anything else
that might help us fixing this issue.
===========================================================
#5  0x000000000040465a in std::default_delete<TTree>::operator()(TTree*) const ()
#6  0x0000000000404317 in std::unique_ptr<TTree, std::default_delete<TTree> >::~unique_ptr() ()
#7  0x0000000000403e45 in writeTree() ()
#8  0x0000000000402e77 in main ()
===========================================================
```

pcanal replied:
This technically because of the:
```
ofile->Close();
```
which deletes the `TTree` object.  This call to `Close` is unnecessary as it is called by the `TFile` destructor and should be replacement by a call to `ofile->Write()` (which makes the call to `tree->Write();` redundant)

atolosadelgado replied:
great, at the end I have changed this part, thanks!

jalopezg-git commented on tutorials/tree/dictionary/writeTree.cpp:
```diff
@@ -0,0 +1,61 @@
+
+   myTree->Write();
+   ofile->Close();
+
+   return;
```
```suggestion
```

jalopezg-git commented on tutorials/tree/dictionary/writeTree.cpp:
```diff
@@ -0,0 +1,61 @@
+   return;
+}
+
+#endif
```
Missing `\n` at the end of last line.

jalopezg-git commented on tutorials/tree/dictionary/readTree.cpp:
```diff
@@ -0,0 +1,52 @@
+void readTree()
+{
+
+   TFile *ifile = TFile::Open("testFile.root", "read");
```
See previous comment.

jalopezg-git commented on tutorials/tree/dictionary/readTree.cpp:
```diff
@@ -0,0 +1,52 @@
+   }
+
+   ifile->Close();
+
+   return;
```
```suggestion
```

jalopezg-git reviewed (changes requested):
Many thanks for the contribution, @atolosadelgado :slightly_smiling_face:!

I have attached some comments though to further improve the code in the PR!

atolosadelgado commented on tutorials/tree/dictionary/CMakeLists.txt:
```diff
@@ -0,0 +1,55 @@
+project(customTreeExample)
+
+# You need to tell CMake where to find the ROOT installation. This can be done in a number of ways:
+#   - ROOT built with classic configure/make use the provided $ROOTSYS/etc/cmake/FindROOT.cmake
```
shall I remove those lines then?

pcanal replied:
yes :)

atolosadelgado commented on tutorials/tree/dictionary/CMakeLists.txt:
```diff
@@ -0,0 +1,55 @@
+#####################################################################################################################
+
+# Jorge Agramunt Ros    @ IFIC(Valencia,Spain)  jorge.agramunt@ific.uv.es
+# Alvaro Tolosa Delgado @ IFIC(Valencia,Spain)  alvaro.tolosa@ific.uv.es
+# Copyright (c) 2016 Jorge Agramunt & Alvaro Tolosa. All rights reserved.
+
+#####################################################################################################################
```
Is there a template text that can replace that header?

pcanal replied:
For C++ files:
```
// Author: Jorge Agramunt Ros IFIC(Valencia,Spain) 07/2023

/*************************************************************************
 * Copyright (C) 1995-2023, Rene Brun and Fons Rademakers.               *
 * All rights reserved.                                                  *
 *                                                                       *
 * For the licensing terms see $ROOTSYS/LICENSE.                         *
 * For the list of contributors see $ROOTSYS/README/CREDITS.             *
 *************************************************************************/
 ```
 For CMake:
```
# Copyright (C) 1995-2023, Rene Brun and Fons Rademakers.
# All rights reserved.
#
# For the licensing terms see $ROOTSYS/LICENSE.
# For the list of contributors see $ROOTSYS/README/CREDITS.
```

atolosadelgado commented:
Hi,

I implemented most of the changes, but I would need some indications for the rest:
1. Is there a standard header for ROOT files?
2. Remove reference in CMakeLists to configure method to configure the ROOT project before building it?
3. `In general, let's also briefly describe here the steps required to read/write user-defined classes, i.e. why ROOT relies on dictionaries for class introspection, how to generate one, etc.` Is a link to the ROOT documentation enough? In addition, I am not expert on the topic, I can not expand your points. 
4. `std::unique_ptr<TTree>` seems to not work. Never used if before, do you have any idea what is failing?

Thank for your time and effort,
Alvaro

pcanal commented on tutorials/tree/dictionary/writeTree.cpp:
```diff
@@ -0,0 +1,55 @@
+
+   myTree->Print();
+
+   myTree->Write();
+   ofile->Close();
```
```suggestion
   ofile->Write();  // This write the files and the TTree
```

pcanal commented on tutorials/tree/dictionary/readTree.cpp:
```diff
@@ -0,0 +1,46 @@
+         std::cerr << "WARNING: entry " << aReader.GetCurrentEntry() << " is empty! " << std::endl;
+   }
+
+   ifile->Close();
```
```suggestion
```
Calling `Close` here is not needed (it is called by the `TFile` destructor).

atolosadelgado replied:
If the Close function is not called, it used to fail (I have experienced that too many times), so I will keep it

pcanal replied:
You should need only a call to `Close` or an implicit or explicit call to the destruction of the `TFile` (which you have by using the `unique_ptr`).  Otherwise there is a bug/deficiency.   This being a tutorials we should use the 'best practices' :)

pcanal replied:
PS/Note: what is more often missing and is compulsory is calling `TFile::Write` when writing a file.

pcanal commented on tutorials/tree/dictionary/CMakeLists.txt:
```diff
@@ -0,0 +1,55 @@
+# You need to tell CMake where to find the ROOT installation. This can be done in a number of ways:
+#   - ROOT built with classic configure/make use the provided $ROOTSYS/etc/cmake/FindROOT.cmake
+#   - ROOT built with CMake. Add in CMAKE_PREFIX_PATH the installation prefix for ROOT
+list(APPEND CMAKE_PREFIX_PATH $ENV{ROOTSYS})
```
`ROOTSYS` is not always defined.

pcanal commented on tutorials/tree/dictionary/CMakeLists.txt:
```diff
@@ -0,0 +1,55 @@
+#---Define useful ROOT functions and macros (e.g. ROOT_GENERATE_DICTIONARY)
+include(${ROOT_USE_FILE})
+
+#---Add include directory of ROOT to the build
+include_directories(${CMAKE_SOURCE_DIR} ${ROOT_INCLUDE_DIRS})
+
+#---Add library directory of ROOT to the build
+link_directories(${ROOT_LIBRARY_DIR})
+
+#---Set Flags
+add_definitions(${ROOT_CXX_FLAGS}  )
```
These should not be needed .. They are meant to be inherited/transferred automatically after adding the `ROOT` libraries as dependencies of a given target.

atolosadelgado commented on tutorials/tree/dictionary/CMakeLists.txt:
```diff
@@ -0,0 +1,55 @@
+target_link_libraries(treeExample ${ROOT_LIBRARIES} data2TreeLib)
+
+# Enable PIC
+set_property(TARGET treeExample PROPERTY POSITION_INDEPENDENT_CODE ON)
```
I removed PIC, the link libraries stay to make the exec

atolosadelgado commented:
Hi,

I have already implemented all the changes proposed by P. Canal. 

The only thing that remains is the proposal by @jalopezg-git of extending the README, but I am afraid that is beyond my knowledge.

Best,
Alvaro

bellenot commented on tutorials/tree/dictionary/CMakeLists.txt:
```diff
@@ -0,0 +1,42 @@
+#   * the class implementation
+add_library(data2TreeLib SHARED data2Tree.cpp G__data2Tree.cxx)
+target_link_libraries(data2TreeLib ${ROOT_LIBRARIES}  ) 
+
```
```suggestion
if(MSVC)
  set_target_properties(data2TreeLib PROPERTIES WINDOWS_EXPORT_ALL_SYMBOLS TRUE)
endif()

```
This is needed on Windows in order to export the symbols and create the `data2TreeLib.lib` file

pcanal commented on tutorials/tree/dictionary/writeTree.cpp:
```diff
@@ -0,0 +1,61 @@
+      return;
+   }
+
+    std::unique_ptr<TTree> myTree = std::make_unique<TTree>("myTree", "");
```
```suggestion
   std::unique_ptr<TTree> myTree = std::make_unique<TTree>("myTree", "");
```

pcanal commented on tutorials/tree/dictionary/readTree.cpp:
```diff
@@ -0,0 +1,51 @@
+void readTree()
+{
+
+   std::unique_ptr<TFile> ifile = std::make_unique<TFile>("testFile.root", "read");
+   if ( nullptr == ifile ) {
+      std::cerr << " File not found " << std::endl;
+      return;
+   }
```
Either:
```suggestion
   std::unique_ptr<TFile> ifile = std::make_unique<TFile>("testFile.root", "read");
   if ( !ifile || ifile->IsZombie() ) {
      std::cerr << " Problem opening the file testFile.root" << std::endl;
      return;
   }
```
Note that here the case (`!file, i.e ifile == nullptr`) will be extremely rare as a constructor can only fails to run if it can no allocate the memory.
```suggestion
   std::unique_ptr<TFile> ifile { TFile::Open("testFile.root", "read") };
   if ( !ifile || ifile->IsZombie() ) {
      std::cerr << " Problem opening the file testFile.root" << std::endl;
      return;
   }
```
which has the advantage of using the syntax that can open remote files.

pcanal commented on tutorials/tree/dictionary/readTree.cpp:
```diff
@@ -0,0 +1,51 @@
+      std::cerr << " File not found " << std::endl;
+      return;
+   }
+   if (!TClass::GetClass(typeid(myDetectorData))->IsLoaded()) {
+      std::cerr << " TClass::GetClass(typeid(myDetectorData))->IsLoaded() == false " << std::endl;
+   }
```
Is this something we expect the user to want to use in all scripts or is it just for debugging purposes?

atolosadelgado replied:
I do not remember why I put that, it was long time ago. I can remove it.

pcanal commented on tutorials/tree/dictionary/readTree.cpp:
```diff
@@ -0,0 +1,51 @@
+   // Loop over the entries of the tree
+   while (aReader.Next()) {
+      if (branch1->time != 0)
+         std::cerr << " -Branch1 : time: " << branch1->time << "\t energy: " << branch1->energy << std::endl;
+      else if (branch2->time != 0)
+         std::cerr << " +Branch2 : time: " << branch2->time << "\t energy: " << branch2->energy << std::endl;
```
```suggestion
         std::cout << " -Branch1 : time: " << branch1->time << "\t energy: " << branch1->energy << std::endl;
      else if (branch2->time != 0)
         std::cout << " +Branch2 : time: " << branch2->time << "\t energy: " << branch2->energy << std::endl;
```
as those are not errors (if I understood correctly) but the 'expected' output.

pcanal commented on tutorials/tree/dictionary/writeTree.cpp:
```diff
@@ -0,0 +1,61 @@
+
+void writeTree()
+{
+   std::unique_ptr<TFile> ofile = std::make_unique<TFile>("testFile.root", "recreate");
```
See https://github.com/root-project/root/pull/13205/files#r1264109492

pcanal commented on tutorials/tree/dictionary/writeTree.cpp:
```diff
@@ -0,0 +1,61 @@
+   myDetectorData obj_for_branch2;
+   myTree->Branch("branch2.", &obj_for_branch2);
+
+   for (int i = 0; i < 10; ++i) {
+      //-- if i is even, fill branch2 and set branch1's entry to zero
+      if (i % 2 == 0) {
+         obj_for_branch1.clear();
+         obj_for_branch2.time = i + 5;
+         obj_for_branch2.energy = 2 * i + 5;
+         obj_for_branch2.detectorID = 3 * i + 5;
+         myTree->Fill();
+
+      }
+      //-- if i is odd, we do the opposite
+      else {
+         obj_for_branch2.clear();
+         obj_for_branch1.time = i + 1;
+         obj_for_branch1.energy = 2 * i + 1;
+         obj_for_branch1.detectorID = 3 * i + 1;
+         myTree->Fill();
+      }
```
```suggestion
   for (int i = 0; i < 10; ++i) {
      //-- if i is even, fill branch2 and set branch1's entry to zero
      if (i % 2 == 0) {
         obj_for_branch1.clear();
         obj_for_branch2.time = i + 5;
         obj_for_branch2.energy = 2 * i + 5;
         obj_for_branch2.detectorID = 3 * i + 5;
      }
      //-- if i is odd, we do the opposite
      else {
         obj_for_branch2.clear();
         obj_for_branch1.time = i + 1;
         obj_for_branch1.energy = 2 * i + 1;
         obj_for_branch1.detectorID = 3 * i + 1;
      }
      myTree->Fill();
```
Coding the common operations once is usually better and more readable.

atolosadelgado commented:
Thanks P. Canal for all your comments, all are already implemented :)

atolosadelgado commented on tutorials/tree/dictionary/writeTree.cpp:
```diff
@@ -0,0 +1,59 @@
+
+   std::unique_ptr<TTree> myTree = std::make_unique<TTree>("myTree", "");
+   myDetectorData obj_for_branch1;
+   myTree->Branch("branch1.", &obj_for_branch1);
```
just out of curiosity, is the dot at the end of the branch name needed?

atolosadelgado replied:
Thank you very much! I think that information is important enough to be part of the documentation :)

pcanal replied:
I see trace of it in https://root.cern.ch/doc/master/classTTree.html

pcanal commented on tutorials/tree/dictionary/main.cpp:
```diff
@@ -0,0 +1,28 @@
+
+int main()
+{
+   std::cerr << "Starting writeTree()..." << std::endl;
```
```suggestion
   std::cout << "Starting writeTree()..." << std::endl;
```

pcanal commented on tutorials/tree/dictionary/main.cpp:
```diff
@@ -0,0 +1,28 @@
+
+int main()
+{
+   std::cerr << "Starting writeTree()..." << std::endl;
+   writeTree();
+   std::cerr << "Starting writeTree()... Done! " << std::endl;
+   std::cerr << "Starting readTree()..." << std::endl;
+   readTree();
+   std::cerr << "Starting readTree()... Done! " << std::endl;
```
```suggestion
   std::cout << "Starting writeTree()..." << std::endl;
   writeTree();
   std::cout << "Starting writeTree()... Done! " << std::endl;
   std::cout << "Starting readTree()..." << std::endl;
   readTree();
   std::cout << "Starting readTree()... Done! " << std::endl;
```

pcanal commented on tutorials/tree/dictionary/readTree.cpp:
```diff
@@ -0,0 +1,48 @@
+      else if (branch2->time != 0)
+         std::cout << " +Branch2 : time: " << branch2->time << "\t energy: " << branch2->energy << std::endl;
+      else
+         std::cout << "WARNING: entry " << aReader.GetCurrentEntry() << " is empty! " << std::endl;
```
```suggestion
         std::cerr << "WARNING: entry " << aReader.GetCurrentEntry() << " is empty! " << std::endl;
```

jalopezg-git commented:
@atolosadelgado let me know where it needs another review by clicking on the Re-request review button in the 'Reviewers' section above.

vepadulano commented on tutorials/tree/dictionary/writeTree.cpp:
```diff
@@ -0,0 +1,59 @@
+void writeTree()
+{
+   std::unique_ptr<TFile> ofile { TFile::Open("testFile.root", "recreate") };
+   if ( nullptr == ofile ) {
+      std::cerr << " File not open." << std::endl;
```
```suggestion
   if (!ofile || ofile->IsZombie()) {
      throw std::runtime_error("Could not open file testFile.root");
```

vepadulano commented on tutorials/tree/dictionary/writeTree.cpp:
```diff
@@ -0,0 +1,59 @@
+
+void writeTree()
+{
+   std::unique_ptr<TFile> ofile { TFile::Open("testFile.root", "recreate") };
```
Very nice using `unique_ptr` in new tutorials, thank you :smile:

vepadulano commented on tutorials/tree/dictionary/writeTree.cpp:
```diff
@@ -0,0 +1,59 @@
+
+   myTree->Print();
+
+   ofile->Write(); // This write the files and the TTree
```
Probably this syntax is clearer for readers and doesn't even need an extra comment
```suggestion
   ofile->WriteObject(&myTree, myTree->GetName());
```

pcanal replied:
@vepadulano The ' ofile->Write();' is necessary (to write the TFile itself) and includes the `WriteObject` because the `TTree` is attached to the file.

vepadulano commented on tutorials/tree/dictionary/readTree.cpp:
```diff
@@ -0,0 +1,48 @@
+
+#include "data2Tree.hpp"
+
+void readTree()
+{
+
+   std::unique_ptr<TFile> ifile { TFile::Open("testFile.root", "read") };
+   if ( !ifile || ifile->IsZombie() )  {
+      std::cerr << "Problem opening the file testFile.root" << std::endl;
+      return;
+   }
+
+   // Create a TTreeReader to read the tree named "myTree"
+   TTreeReader aReader("myTree", ifile.get() );
+
+   // Create TTreeReaderValues for the branches "branch1" and "branch2"
+   TTreeReaderValue<myDetectorData> branch1(aReader, "branch1.");
+   TTreeReaderValue<myDetectorData> branch2(aReader, "branch2.");
+
+   // Loop over the entries of the tree
+   while (aReader.Next()) {
+      if (branch1->time != 0)
+         std::cout << " -Branch1 : time: " << branch1->time << "\t energy: " << branch1->energy << std::endl;
+      else if (branch2->time != 0)
+         std::cout << " +Branch2 : time: " << branch2->time << "\t energy: " << branch2->energy << std::endl;
+      else
+         std::cerr << "WARNING: entry " << aReader.GetCurrentEntry() << " is empty! " << std::endl;
+   }
+
+}
+
```
I understand that for the writing part of the tutorial we want to show usage of bare TTree, but for the processing I believe we should really only show RDataFrame at this point. I don't see any reason to use TTreeReader as an analysis interface here.

I would suggest reframing this part of the tutorial with RDataFrame. Something akin to this should be a good starting point
```
ROOT::RDataFrame df{"myTree", "testFile.root"};
df.Display({"branch1.time","branch1.energy","branch2.time","branch2.energy"}, /*nRows*/ 10)->Print();
```

vepadulano commented on tutorials/tree/dictionary/readTree.cpp:
```diff
@@ -0,0 +1,48 @@
+// Author: Alvaro Tolosa-Delgado CERN 07/2023
```
I believe that `readTree.cpp` should be renamed to `readTree.cxx` for symmetry with practically every other part of ROOT, similarly for `.hpp` to `.hxx`

vepadulano reviewed (changes requested):
Dear @atolosadelgado , thank you so much for this tutorial! I believe it makes a great addition and clarifies a lot of the workflow of generating a dictionary for usage of TTree. I left a couple of minor comments, together with a strong suggestion of moving away from TTreeReader into RDataFrame for the reading example.

atolosadelgado commented:
Hi @jalopezg-git et al :)

Thank you for your comments and suggestions, I have learnt new things in this PR.

In the last commits I implemented the changes suggested by @vepadulano. Note that I have added a new file with the two lines that print the contents of the TTree using a RDF.

The topics that would remain open are:
* Extension of README, as stated [here](https://github.com/root-project/root/pull/13205#discussion_r1259443075). I am sorry but that is beyond my knowledge. Please feel free to complement the README of this tutorial or the ROOT documentation.
* The dot at the end of the branch name is not documented, but it is discussed in some forum threads. I would suggest to add the comment from P. Canal [here](https://github.com/root-project/root/pull/13205#discussion_r1265655686) for sake of clarity.

Thank you for your time.

Best,
Alvaro

jalopezg-git reviewed (changes requested):
Almost there :+1:!  I have included some suggestions to apply before merging!

Also re-requesting review to @vepadulano and @bellenot.

atolosadelgado commented on tutorials/tree/dictionary/writeTree.cxx:
```diff
@@ -0,0 +1,58 @@
+   for (int i = 0; i < 10; ++i) {
+      //-- if i is even, fill branch2 and set branch1's entry to zero
+      if (i % 2 == 0) {
+         obj_for_branch1.clear();
+         obj_for_branch2.time = i + 5;
+         obj_for_branch2.energy = 2 * i + 5;
+         obj_for_branch2.detectorID = 3 * i + 5;
+      }
+      //-- if i is odd, we do the opposite
+      else {
+         obj_for_branch2.clear();
+         obj_for_branch1.time = i + 1;
+         obj_for_branch1.energy = 2 * i + 1;
+         obj_for_branch1.detectorID = 3 * i + 1;
```
I would keep the code as it is because the main goal is learning how to use dictionaries.

atolosadelgado commented:
> Almost there +1! I have included some suggestions to apply before merging! I'll re-request review to @vepadulano and @bellenot.
> 
> Also note:
> 
>     * I didn't see the changes suggested in [this comment](https://github.com/root-project/root/pull/13205#discussion_r1265655686) by @pcanal.
> 
>     * `clang-format` suggestions will have to be applied before merging.

Hi Javier, thank you for your comments. 

If I understood properly, the comment by P. Canal about the dot at the end of the branch name means that it is mandatory in this case to avoid degeneracy, and a good practice in general. For that reason the branch names in this tutorial end with a dot.

Regarding the README file, I have added your suggestion about the linkdef file. Would you add something else?

The files *.c* and *.h* now follow the clang-format style.

jalopezg-git commented on tutorials/tree/dictionary/writeTree.cxx:
```diff
@@ -0,0 +1,56 @@
+
+   std::unique_ptr<TTree> myTree = std::make_unique<TTree>("myTree", "");
+   myDetectorData obj_for_branch1;
+   myTree->Branch("branch1.", &obj_for_branch1);
```
I am still missing here a brief comment, perhaps 1 or 2 lines, explaining of the trailing dot (see [@pcanal's comment](https://github.com/root-project/root/pull/13205#discussion_r1265655686)).

jalopezg-git reviewed (commented):
See comment.  Also, I guess it builds and works out of the box, right? (didn't test it myself)

atolosadelgado commented:
> See comment. Also, I guess it builds and works out of the box, right? (didn't test it myself)
> 
> > Regarding the README file, I have added your suggestion about the linkdef file. Would you add something else?
> 
> As a first version, I think it's okay as-is, but let's also see what @vepadulano and @pcanal think.

Hi Javier,

I have added some comments in the writeTree.cpp about the final dot in the branch name. In addition, I have added 2 sentences to the README.

Thank you for your patience.
Alvaro

jalopezg-git reviewed (commented):
I think it's essentially good to go :+1:; I stil have to try that it builds and compiles without any issue though!

@vepadulano, @pcanal could you take another look too?  If you also think it's ready, we could merge even this week.

jalopezg-git commented:
@phsft-bot build

vepadulano reviewed (approved):
Thanks again for this nice work! I think it's a good demonstration of the dictionary capabilities. I left a last minor comment to make a clarification in the doc.

vepadulano commented:
One other suggestion that slipped through my review. 45 commits is probably too much for this PR and they should be squashed into one or a few more.

atolosadelgado commented on tutorials/tree/dictionary/CMakeLists.txt:
```diff
@@ -0,0 +1,47 @@
+#---Locate the ROOT package and defines a number of variables (e.g. ROOT_INCLUDE_DIRS)
+find_package(ROOT REQUIRED COMPONENTS Tree TreePlayer ROOTDataFrame)
+
+#---Define useful ROOT functions and macros (e.g. ROOT_GENERATE_DICTIONARY)
+include(${ROOT_USE_FILE})
```
do you think a reference to this post would be useful?
https://root-forum.cern.ch/t/how-to-integrate-root-into-my-project-with-cmake/37175
I can copypaste Couet explanation  of that point

jalopezg-git commented:
> One other suggestion that slipped through my review. 45 commits is probably too much for this PR and they should be squashed into one or a few more.

Indeed, we could squash on merge into a single commit; what do you think, @atolosadelgado?

jalopezg-git commented:
@phsft-bot build

jalopezg-git reviewed (commented):
Many thanks for the contribution!  I tried the tutorial and it works without any issue :+1:!  The PR will be merged after the CI test results are available.

atolosadelgado commented on tutorials/tree/dictionary/CMakeLists.txt:
```diff
@@ -0,0 +1,49 @@
+#---Include a CMake module which makes use of the previous variables and loads modules 
+#   with useful macros or functions such as ROOT_GENERATE_DICTIONARY
+#   For further details: https://root-forum.cern.ch/t/how-to-integrate-root-into-my-project-with-cmake/37175
+include(${ROOT_USE_FILE})
```
Thank you! I think those tips are worth to be included as part of the official  ROOT documentation :)
